### PR TITLE
[JENKINS-50164] Remove UI option to modify WS and builds locations

### PR DESCRIPTION
**Pull request by batmat**
Introduce system properties in lieu of the previous way using the UI for this.
Still leave the UI in readonly mode to ease checking configuration for admins.

Tests have been written to check using those system properties would work, even on an existing setup.

A *warning* log is now outputted when these system properties are used and changing the previous value, so that not only the help text warns about the fact there's no data migration.

See [JENKINS-50164](https://issues.jenkins-ci.org/browse/JENKINS-50164).

### TODO before/just after merge:

~~* [x] document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published. Requested by @daniel-beck  https://github.com/jenkins-infra/jenkins.io/pull/1510~~
* [x] Update the [sysprop standard page](https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties) with new syprops

### Proposed changelog entries

* JENKINS-50164: Removed the option to modify _Build Record Root Directory_ and _Workspace Root Directory_ from the User Interface directly. Those are now only modifiable using system properties, refer to the [standard page for these values](https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties)  **TODO: write docs there once/if merged**. Remember that no data migration is handled when using those options on an existing setup. 

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: And need to `save()`, I suspect.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: `@see #workspaceDir`

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jglick) [inline core/src/main/java/jenkins/model/Jenkins.java]: I do not think these constants need to be `public`—we do not expect plugin Java code to refer to them.

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (jglick) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: stray space

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (jglick) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: Setting the `readonly` attribute will block a web browser from editing this field, but `doConfigSubmit` is continuing to accept modifications from HTTP posts! I think you want to delete those two lines as well.

I would really just suggest deleting this `advanced` block, and print a warning to the system log if there is an override which is not matched by the system property.

File: test/src/test/java/jenkins/model/JenkinsInitializationTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/JenkinsInitializationTest.java]: Use `then`, which takes a SAM that can be implemented via lambda.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: done, see below

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: done

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (batmat) [reply core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: done it all

File: test/src/test/java/jenkins/model/JenkinsInitializationTest.java
Comment (batmat) [reply test/src/test/java/jenkins/model/JenkinsInitializationTest.java]: done

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: fixed

**Comment by batmat** [comment]
@jglick I believe I addressed all your comments.

Only thing not done yet is checking how it works with https://issues.jenkins-ci.org/browse/JENKINS-21942 as requested in https://issues.jenkins-ci.org/browse/JENKINS-50164?focusedCommentId=332902&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-332902

**Comment by batmat** [comment]
Build succeeded, reviews are welcome. Thanks!

**Comment by jtnord** [comment]
I am sceptical about this - esp because of a JIRA I am currently trying to track down - and also as this should be managed as CaC and not a System property.

Because as was said to me on a different PR No NOT another randomly hidden system property that alters behaviour / state.

**Comment by daniel-beck** [comment]
@jtnord What alternative to having a UI option (which clearly leads to broken behavior) do you propose?

**Comment by jtnord** [comment]
@daniel-beck 

> also as this should be managed as CaC and not a System property.

The UI option can lead to broken behaviour just as much as a System property. 
At least in the UI you can do `FormValidation` such that they can be warned before they submit...

**Comment by daniel-beck** [comment]
@jtnord 

> The UI option can lead to broken behaviour just as much as a System property.

Not really, as it cannot easily be changed while Jenkins is running ([and I'm trying to prevent CasC doing something dumb in this comment](https://github.com/jenkinsci/configuration-as-code-plugin/issues/151#issuecomment-374012341)). Pulling the rug out from under builds in memory is pretty much unique to the UI option. (Also, no sane way around changing it via the UI, unless you really like XML).

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: seems scary as this is being called in a thread that is performing a load

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (jtnord) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: Seems to be missing a test for the following:

* Customer build/workspace are set in existing Jenkins config (from version without this change)
* Jenkins is upgrades
* Jenkins config is saved
* configuration should not be reset - existing custom values should be retained until they set the system property

Because otherwise people are going to not do that step and mess things up.

And this implies there is no way to go back to default - once you have set the system property if you want to revert to the defaults you need to know what they are and set the system property to it (removing the system property should not change things)...

**Comment by daniel-beck** [comment]
It is unclear why #3360 was closed instead of updated, but given @jtnord's comments it seems quite a few of the review comments there are still relevant here.

**Comment by jtnord** [comment]
By broken I mean a user setting BUILD_ROOT=ITEM_ROOT.  [JENKINS-50335](https://issues.jenkins-ci.org/browse/JENKINS-50335)

**Comment by jtnord** [comment]
A job can configure a custom workspace, not sure how that overlaps with this PR.

**Comment by batmat** [comment]
> A job can configure a custom workspace, not sure how that overlaps with this PR.

IMO it does not. Or I mean, yes, this seems like an issue. But if so, was/is already existing so should be handled in a separate PR anyway.

**Comment by batmat** [comment]
> It is unclear why #3360 was closed instead of updated

Yes. I hesitated, and went for a brand new PR as I didn't reuse anything from the previous one. The alternative was to force push, could have done that. (The 3rd alternative, to just add more commits and keep the outdated ones I really do not like, in such case this leads IMO lead to only messing things to use `git blame` afterwards).
 
> but given @jtnord's comments it seems quite a few of the review comments there are still relevant here.

Care to be more specific? I did spend a bit of time here to actually write migration tests, so apart from possibly addressing the specific use case noted by @jtnord, I'm not sure I see *exactly* what use case you think is not covered.

**Comment by batmat** [comment]
@jtnord just pushed https://github.com/jenkinsci/jenkins/pull/3364/commits/774388598e1bf861330c29470484d7957d86fe69 that adds a test showing the use case you described in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r177042462 is tested. Could you please have another look?

@daniel-beck I've checked the other comments here and in #3360 and FWICT I've covered all of them by tests or answers.

Thanks a lot.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: Recommended by @jglick in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r176423228

I'm not totally sure myself of the criticality of the risk, if there is one.

**Review by jtnord — DISMISSED** [review]

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: So we're removing the readonly UI after all?

**Review by daniel-beck — DISMISSED** [review]
This is dropping all the elaborate form validation I wrote to prevent terrible values, which seems much more likely with users accidentally not, or double, quoting the command args on the shell (i.e. `/builds` and `/workspace/` respectively for the 'default' values since the variables evaluate to the empty string).

What currently is [form validation](https://github.com/batmat/jenkins/blob/2c1fdd26990b417a4e57fdaaa7fb30ddc2a9be17/core/src/main/java/jenkins/model/Jenkins.java#L2140...L2178) should become errors preventing Jenkins startup.

Also, that unused code needs to then be removed.

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (batmat) [reply core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: @daniel-beck I finally took the path Jesse advised in https://github.com/jenkinsci/jenkins/pull/3364#discussion_r176424745

Comment (daniel-beck) [reply core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: Preventing manipulated browser form submissions and not showing the current configuration are not necessarily the same thing.

This could easily be a table showing the current values and perhaps providing example paths.

**Comment by batmat** [comment]
@daniel-beck I believe I addressed your comment in the last commit.

**Comment by batmat** [comment]
@reviewbybees for more eyes.

@jtnord could you possibly please review again the addition?

@daniel-beck do you think you can have another look? 

Thanks

**Review by daniel-beck — APPROVED** [review]
In isolation, this change is now OK. I tested the behavior and what was form validation is now boot failures.

Since we're removing documentation from Jenkins, make sure to document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published.

Also, make sure the PR comment and changelog message are updated.

**Comment by daniel-beck** [comment]
Would strongly prefer more reviews than just mine for this change.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: this would result in an illegal filename on windows as `:` is a reserved character - how are the tests passing??

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: this method is utterly confusing

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: @jtnord no idea. Can check, but this test is a cut-n-paste from before, and looking around here, it does not seem it was in a test using `assume`. Checking on CI side.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [inline core/src/main/java/jenkins/model/Jenkins.java]: useless javadoc.  what is it checking for - embedding bitcoins?

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (jtnord) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: so there are no bad values for windows??  this seems to be unix specific without an assume and may fail but likely not for the expected reason.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (jtnord) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: NIT: it is much nicer to use `assumeFalse(" Default Windows lifecycle does not support restart.", Function.isWindows())` as then the reason for the skipped test will be reported.

**Review by jtnord — COMMENTED** [review]
Seems ok apart from the `checkRawBuildsDir` method which is thoroughly confusing to me.

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (jtnord) [reply core/src/main/java/jenkins/model/Jenkins.java]: I found it really hard to even grok what this whole method was trying to do. 
if it is cut and paste then _shrug_

Comment (daniel-beck) [reply core/src/main/java/jenkins/model/Jenkins.java]: We're trying to figure out whether there would be legal file names for a given item name (without having an item). This particular part implements the `ITEM_FULLNAME` vs. `ITEM_FULL_NAME` compatibility check.

Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: So, didn't dig into _why_ it works, but this test, from `JenkinsDescriptorTest` **is** currently successful in CI on master: https://ci.jenkins.io/job/Core/job/jenkins/job/master/867/testReport/jenkins.model/JenkinsDescriptorTest/Windows___Windows_Publishing___testBuildDirValidation/

image

Comment (daniel-beck) [reply core/src/main/java/jenkins/model/Jenkins.java]: @batmat There is no test for `ITEM_FULLNAME` due to the platform dependency, so this doesn't even happen during the test. This used to be basic form validation, so I didn't bother.

What's the problem here anyway? If the user wants `ITEM_FULLNAME`, that's only OK if the file system can handle colons. That's all this check does. Otherwise the error message will tell them to use `ITEM_FULL_NAME`, which escapes colons.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (batmat) [reply test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: Fixed

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (batmat) [reply core/src/main/java/jenkins/model/Jenkins.java]: Fixed

**Comment by batmat** [comment]
I fixed some of your comments @jtnord. Thanks for the review!

To /all I plan to merge this before the next release, i.e. on Sunday max if nobody objects in the meantime.

Thanks!

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (jglick) [reply core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: IMO: users should not be customizing this. If they do, they presumably have some reason and will be able to figure out example paths for themselves. If they are not customizing this, then they should have no reason to care. And the workspace root is only relevant for on-master builds, which you should never do to begin with.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: No `@since` tag needed on a test.

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (jglick) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: The lambda takes a parameter. And `getInstance` is `static`. Conventionally

```java
story.then(r -> {
    assertTrue(r.jenkins.isDefaultBuildDir());
```

**Review by jglick — APPROVED** [review]
Not sure I understand all the implications, but removing a UI option is generally a good thing, and I am assuming you are testing this.

File: core/src/main/resources/jenkins/model/Jenkins/configure.jelly
Comment (daniel-beck) [reply core/src/main/resources/jenkins/model/Jenkins/configure.jelly]: @jglick 

> And the workspace root is only relevant for on-master builds, which you should never do to begin with.

Jenkins does it for you in matrix parents, pipelines, etc.

ci.j.io gets the path sanitization warning from `jenkins.branch.WorkspaceLocatorImpl locate` regularly due to the legacy setting, and there are no builds executing on master.

**Comment by daniel-beck** [comment]
On-holding this until the following is addressed (or at least gets a reasonable response):

> Since we're removing documentation from Jenkins, make sure to document the new behavior, e.g. on the wiki or in the Jenkins handbook. This should be ready by the time the weekly changelog is published.
> 
> Also, make sure the PR comment and changelog message are updated.

**Comment by batmat** [comment]
Working on the _on-holding_ reason now, hopefully to propose something very soon.

To loop back with a discussion on IRC with Daniel, I'm generally against squashing. So please just do not squash this one if you merge it. Thanks!

**Comment by batmat** [comment]
Just filed the PR linked above to _on-hold_ this. Please review so that we can move forward here.

Thanks!

**Comment by batmat** [comment]
Updated https://wiki.jenkins.io/display/JENKINS/Features+controlled+by+system+properties?blah=eeek

**Comment by batmat** [comment]
Going to merge this in the next 10 minutes if nobody objects. The PR was already approved, and I've fixed the reason for putting it On-Hold.

**Comment by batmat** [comment]
For the record, the PR was built green previously. The current failure seems like an infra issue.

image

image

**Comment by KostyaSha** [comment]
Am i correctly understand:
Jenkins.java: `    private static final String DEFAULT_BUILDS_DIR = "${ITEM_ROOTDIR}/builds";` is a default value
Then `JenkinsRule` calls `Jenkins.loadTasks()` -> `Jenkins.loadConfig();`, by default `config.xml` is missing, so `checkRawBuildsDir(buildsDir)` gets `checkRawBuildsDir("${ITEM_ROOTDIR}/builds")` which expands `...../jobs/doCheckRawBuildsDir-Marker$foo/builds` and failed because it's not an absolute path?

**Comment by KostyaSha** [comment]
@jglick @batmat @daniel-beck @oleg-nenashev ?

File: core/src/main/java/jenkins/model/Jenkins.java
Comment (KostyaSha) [inline core/src/main/java/jenkins/model/Jenkins.java]: should it be covered with a test cases?

File: test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java
Comment (KostyaSha) [inline test/src/test/java/jenkins/model/JenkinsBuildsAndWorkspacesDirectoriesTest.java]: so is it good or bad?

**Comment by jglick** [comment]
> when empty Jenkins is running does it hit into wrong default

[JENKINS-21942](https://issues.jenkins-ci.org/browse/JENKINS-21942)

**Pull request merged**
